momentumopt/python/momentumopt/kinoptpy/second_order_ik.py

```diff
@@ -0,0 +1,262 @@
+import numpy as np
+import matplotlib.pyplot as plt
+
+import pinocchio as pin
+from pinocchio import RobotWrapper
+from pinocchio.utils import zero
+
+import scipy.linalg
+
+from scipy.interpolate import CubicSpline
+
+from momentumopt.kinoptpy.qp import QpSolver
+
+
+class SecondOrderInverseKinematics(object):
+    def __init__(self, model, endeff_frame_names):
+
+        def getFrameId(name):
+            idx = self.robot.model.getFrameId(name)
+            if idx == len(self.robot.model.frames):
+                raise Exception('Unknown frame name: {}'.format(name))
+            return idx
+
+        self.robot = RobotWrapper(model)
+        self.robot_mass = sum([i.mass for i in self.robot.model.inertias[1:]])
+        self.base_id = getFrameId('base_link')
+        self.endeff_frame_names = endeff_frame_names
+        self.endeff_ids = [getFrameId(name) for name in endeff_frame_names]
+        self.is_init_time = True
+
+        self.ne = len(self.endeff_ids)
+        self.nv = self.robot.model.nv
+
+        # Tracking weights
+        self.w_endeff_tracking = 10**5
+        self.w_endeff_contact = 10**5
+        self.w_lin_mom_tracking = 100.0
+        self.w_ang_mom_tracking = 10.0
+        self.w_joint_regularization = 0.1
+
+        # P and D gains for tracking the position
+        self.p_endeff_tracking = 10000.
+        self.d_endeff_tracking = 200
+
+        self.p_com_tracking = 100.
+        self.p_orient_tracking = 10.
+        self.d_orient_tracking = 1.
+        self.mom_tracking = 10. * np.array([1., 1., 1., .01, .01, .01])
+
+        self.p_joint_regularization = 1.
+        self.d_joint_regularization = .5
+
+        self.desired_acceleration = np.zeros(((self.ne + 2) * 3 + (self.nv - 6), ))
+
+
+        # Allocate space for the jacobian and desired velocities.
+        # Using two entires for the linear and angular velocity of the base.
+        # (self.nv - 6) is the number of joints for posture regularization
+        self.J = np.zeros(((self.ne + 2) * 3 + (self.nv - 6), self.nv))
+        self.drift_terms = np.zeros_like(self.desired_acceleration) #i.e. dJ * dq
+        self.measured_velocities = np.zeros((self.J.shape[0], )) # i.e. J * dq
+
+        self.use_hierarchy = False
+        self.qp_solver = QpSolver()
+
+    def framesPos(self, frames):
+        """
+            puts the translations of the list of frames in a len(frames) x 3 array
+        """
+        return np.vstack([self.robot.data.oMf[idx].translation for idx in frames]).reshape([len(frames),3])
+
+
+    def update_des_acc(self, q, dq, com_ref, orien_ref, mom_ref, dmom_ref,
+                        endeff_pos_ref, endeff_vel_ref, endeff_acc_ref,
+                        joint_regularization_ref):
+
+        measured_op_space_velocities = self.J @ dq
```

Comment:
this works, but there might be efficient ways to do that in pinocchio?
We can let the optimization for another PR